theano/contrib/mkl/mkl_ndarray.c

```diff
@@ -0,0 +1,1361 @@
+    } else {
+        return PyInt_FromLong(self->nd);
+    }
+}
+
+
+/*
+ * Get size info of a MKLNdarray instance.
+ *
+ * Register in MKLNdarray_getset
+ *
+ * Return a integer number.
+ *
+ */
+static PyObject*
+MKLNdarray_get_size(MKLNdarray *self, void *closure) {
+    size_t total_element = 1;
+    if (self == NULL) {
+        PyErr_SetString(PyExc_RuntimeError,
+                        "MKLNdarray_get_size: input MKLNdarray* self is NULL");
+        return NULL;
+    }
+
+    if (self->nd <= 0) {
+        total_element = 0;
+    } else {
+        for (int i = 0; i < self->nd; i++) {
+            total_element *= self->user_structure[i];
+        }
+    }
+    return PyInt_FromLong(total_element);
+}
+
+
+/*
+ * Get base info of a MKLNdarray instance.
+ *
+ * Register in MKLNdarray_getset
+ *
+ * Return a PyObject.
+ *
+ */
+static PyObject*
+MKLNdarray_get_base(MKLNdarray *self, void *closure) {
+    if (self == NULL) {
+        PyErr_SetString(PyExc_RuntimeError,
+                        "MKLNdarray_get_base: input MKLNdarray* self is NULL");
+        return NULL;
+    }
+
+    PyObject * base = self->base;
+    if (!base) {
+        base = Py_None;
+    }
+    Py_INCREF(base);
+    return base;
+}
+
+
+/*
+ * Create a PyArrayObject from a MKLNdarray.
+ */
+PyObject* MKLNdarray_CreateArrayObj(MKLNdarray *self) {
+    if (self == NULL) {
+        PyErr_SetString(PyExc_RuntimeError,
+                        "MKLNdarray_CreateArrayObj: input MKLNdarray* self is NULL");
+        return NULL;
+    }
+
+    if (self->nd < 0 ||
+        self->private_data == NULL ||
+        self->private_layout == NULL) {
+        PyErr_SetString(PyExc_RuntimeError,
+                        "MKLNdarray_CreateArrayObj: Can't convert from an uninitialized MKLNdarray");
+        return NULL;
+    }
+
+    npy_intp npydims[MAX_NDIM] = {0};
+    for (int i = 0; i < self->nd; i++) {
+        npydims[i] = (npy_intp)self->user_structure[i];
+    }
+
+    PyArrayObject* rval = NULL;
+    if (self->dtype == MKL_FLOAT64) {
+        // float64
+        rval = (PyArrayObject*)PyArray_SimpleNew(self->nd, npydims, NPY_FLOAT64);
+    } else {
+        // float32
+        rval = (PyArrayObject*)PyArray_SimpleNew(self->nd, npydims, NPY_FLOAT32);
+    }
+
+    if (!rval) {
+        return NULL;
+    }
+
+    void* rval_data = PyArray_DATA(rval);
+    dnnLayout_t layout_user = NULL;
+    int status = -1;
+    dnnPrimitive_t primitive = NULL;
+
+    size_t mkl_size[MAX_NDIM] = {0};
+    size_t mkl_stride[MAX_NDIM] = {0};
+
+    // nchw -> whcn
+    for (int i = 0; i < self->nd; i++) {
+        mkl_size[i] = (MKLNdarray_DIMS(self))[self->nd - i - 1];
+        mkl_stride[i] = (MKLNdarray_STRIDES(self))[self->nd - i -1];
+    }
+
+    if (self->dtype == MKL_FLOAT64) { // float64
+        status = dnnLayoutCreate_F64(&layout_user,
+                                     self->nd,
+                                     mkl_size,
+                                     mkl_stride);
```

Comment:
Yes, you're right about the conversion. But we can't create the conversion primitive for two same layout, so we can't call the convert function (which requires conversion primitive as one of its parameters) to do the simple copy work.

No matter we convert private layout to user layout or vice verse, private layout is always an internal description for mkl object, but user layout is built from Numpy Ndarray actually.

For example, wehen convert private layout to user layout, so the destination layout is layout_user. Although the information we used to build the layout_user is `mkl_size` and `mkl_stride` (and they're stored in user_structure of mkl object), they're acquired from Numpy Ndarray (`dims` and `strides` respectively), the only difference between `mkl_size` and `dims` is element order, `mkl_size` use `WHCN` order, while `dims` use `NCHW` order, same for strides.

So mkl_strides should be ok in this case.